### PR TITLE
Testing - Unit - rspec

### PR DESCRIPTION
**Pull request by ErinBuckley**
@klynton  ready

### CODE REVIEW

**Comment by ErinBuckley** [comment]
@klynton ready for review

File: courses/puppet/rspec_puppet_unit_tests.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
+# Testing Puppet: Unit Tests
+
+As your codebase grows in complexity, changing one part of the code without affecting other parts becomes exponentially more difficult. Unit tests are a method for ensuring that individual parts of your codebase continue to work as intended while other parts are improved and refactored as needed.
+
+At the end of this course you will be able to:
+
+* Explain the purpose of unit testing.
+* Describe methods for using spec-puppet to test.
```
Comment (klynton) [inline courses/puppet/rspec_puppet_unit_tests.md]: rspec-puppet

File: courses/puppet/rspec_puppet_unit_tests.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+In this course we look at how to write unit tests for your Puppet modules using the rspec testing library. But first - a brief review of the structure of a Puppet module. 
+
+### slide 3 - Module Review
+
+Modules are self-contained bundles of code and data. On disk, a module is a directory tree with a specific, predictable structure. The module name is the outermost directory’s name. The manifests directory contains all of the manifests in the module, including an `init.pp` manifest where the class name matches the module name. Other manifest file names map to the names of the classes they contain. A well-formed Puppet module implements each class in a separate file in the manifests directory. Although none of the directories in the module tree are mandatory, for the purposes of testing, you should have an examples sub-directory that contains the test examples that you write for smoke testing your class declarations and defined types. And you should have a spec directory where you put spec and server spec tests, including unit tests, that you use to test plugins in the lib directory.
```
Comment (klynton) [inline courses/puppet/rspec_puppet_unit_tests.md]: "Although none of the directories in the module tree are mandatory" should be changed to "Although having an examples directory isn't mandatory"

The manifests directory is required for your classes, inside of a module.

File: courses/puppet/rspec_puppet_unit_tests.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+You'll also want to check that the catalog contains resources. To check whether a resource exists you use the contain matcher and the resource type for each resource type and helper that (you want to test for. This example  example is checking to see whether an s s h service has been declared.
+
+### slide 13 - Specified Resource Attributes
+
+To validate that resources have specified attributes, you use the contain matcher and the resource type and add the with or without chains and the attribute that you want to check for. In the first example we are checking tfor a file with a specific owner, root. And in the second example we are checking for a file that has the mode attribute undefined.
+
+### slide 14 - Resource relationships
+
+You can also use matchers to validate that resources have relationships set. You can test the relationships between the resources in your catalog regardless of how the jrelationship is defined. IN otherwords, you can define relationships with the metaparameters require, before, notify, and subscrioe, or with chaining arrows. Once again you use the should contain type matcher and add the relationship matcher
+
```
Comment (klynton) [inline courses/puppet/rspec_puppet_unit_tests.md]: jrelationship should be relationship.

subscrioe should be subscribe.

File: courses/puppet/rspec_puppet_unit_tests.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+### slide 18 - Install rspec-puppet
+
+If you used the Puppet module generate to create your module, Puppet automatically created spec directory for you. However, you still need to install the Ruby gems. To install the rspec-puppet gem, type the command `/opt/puppet/bin/gem install rspec-puppet`. Then, to set up tests, install the puppetlabs_spec_helper gem. Type the command `/opt/puppet/bin/gem install puppetlabs_spec_helper`. You may notice that we use Puppet's vendored Ruby, instead of the system Ruby, to install. Although not always the best practice, it happens to be the simplest method of making the proper libraries and gems available for rpsec-puppet to load. You should only install a gem into Puppet's vendoored path if that gem needs to work with Puppet in some way. 
+
```
Comment (klynton) [inline courses/puppet/rspec_puppet_unit_tests.md]: With today's release we'll need to change the path to `/opt/puppetlabs/puppet/bin/gem` in both instances.

**Comment by klynton** [comment]
@ErinBuckley reviewed.

**Pull request merged**